### PR TITLE
allow M to be a Parameter in the huber function

### PR DESCRIPTION
**Pull request by syanga**
## Description
This change allows the M parameter in the huber function to be a parameter for DPP.

## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [x] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [x] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [x] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

File: cvxpy/tests/test_atoms.py [hunk 1]
```diff
@@ -840,6 +840,19 @@ def test_huber(self) -> None:
         self.assertTrue(copy.args[0] is self.y)
         self.assertEqual(copy.get_data()[0].value, atom.get_data()[0].value)
 
+        # Test usage of M as parameter
+        M = Parameter(nonneg=True)
+        expr = cp.huber(1.0, M)
```
Comment (phschiele) [inline cvxpy/tests/test_atoms.py:845]: Probably we should also add a test case for something like `cp.huber(1.0,  2 * M)` (and decide if we want that to work or not).

File: cvxpy/atoms/elementwise/huber.py [hunk 2]
```diff
@@ -41,68 +42,62 @@ class huber(Elementwise):
...
     x : Expression
         The expression to which the huber function will be applied.
-    M : Constant
+    M : Constant or Parameter
         A scalar constant.
     """
 
     def __init__(self, x, M: int = 1) -> None:
-        self.M = self.cast_to_const(M)
+        self.M = M if isinstance(M, Parameter) else self.cast_to_const(M)
```
Comment (SteveDiamond) [inline cvxpy/atoms/elementwise/huber.py:50]: This change is not necessary.

File: cvxpy/atoms/elementwise/huber.py [hunk 3]
```diff
@@ -41,68 +42,62 @@ class huber(Elementwise):
...
         super(huber, self).__init__(x)
 
+    def parameters(self):
```
Comment (SteveDiamond) [inline cvxpy/atoms/elementwise/huber.py:53]: This function is not necessary I believe.

Comment (SteveDiamond) [inline cvxpy/atoms/elementwise/huber.py:53]: This should be more like ```return super().parameters() + self.M.parameters()```. Your code doesn't handle the case where M is an expression that is parameterized.

File: cvxpy/tests/test_atoms.py (hunk 1 above)
Comment (PTNobel) [reply cvxpy/tests/test_atoms.py:845]: I think we do want that to work. Do we have any other situation where an affine expression of a parameter wouldn't work?

Comment (syanga) [reply cvxpy/tests/test_atoms.py:845]: I added a test for affine expressions.

File: cvxpy/atoms/elementwise/huber.py [hunk 4]
```diff
@@ -41,68 +42,62 @@ class huber(Elementwise):
...
         A scalar constant.
     """
 
     def __init__(self, x, M: int = 1) -> None:
         self.M = self.cast_to_const(M)
         super(huber, self).__init__(x)
 
+    def parameters(self):
+        """If M is a Parameter, include it in the list of Parameters"""
+        if isinstance(self.M, Parameter):
```
Comment (PTNobel) [inline cvxpy/atoms/elementwise/huber.py:55]: What if M is a parametrized expression? I don't see why we have an if clause here

**Review by PTNobel — CHANGES_REQUESTED** [review]

**Review by PTNobel — APPROVED** [review]

File: cvxpy/tests/test_atoms.py [hunk 5]
```diff
@@ -840,6 +840,27 @@ def test_huber(self) -> None:
...
+        self.assertAlmostEqual(expr.value, 1.0)
+
+        # Test M as affine expression of parameter
+        expr = cp.huber(1.0, 0.25 * M + 0.25)
+        M.value = 1
+        self.assertAlmostEqual(expr.value, 0.75)
+        M.value = 0.5
+        self.assertAlmostEqual(expr.value, 0.609375)
+        M.value = 1
+        self.assertAlmostEqual(expr.value, 0.75)
```
Comment (SteveDiamond) [inline cvxpy/tests/test_atoms.py:862]: Could we test that huber with parameterized M works properly with DPP? We have had some recent correctness bugs related to DPP so it is important to investigate/cover with tests.

**Review by SteveDiamond — APPROVED** [review]
Thanks!!!

**Comment by SteveDiamond** [comment]
@syanga nice job!

**Pull request merged**